calico/openstack/mech_calico.py

```diff
@@ -591,6 +727,23 @@ def periodic_resync_thread(self, expected_epoch):
         else:
             LOG.warning("Periodic resync thread exiting.")
 
+    def complete_resync(self, context):
+        """Go through complete resync.
+        """
+        if not self.in_resync:
```

Comment:
I don't think this is correct.  There's no guarantee that because this resync is started, it's safe to ignore another resync that comes in.  You could make it more correct by using a proper lock so the two threads would queue up to enter this method.

However, I'm very uneasy about having two threads with overlapping responsibilities.  It doesn't seem right that your new thread calls resync_enpoints via this method, for example.

Is there anything wrong with leaving the two threads as "ships in the night", responsible for different subtrees?